File: services/search-service/src/test/java/com/visor/school/search/service/SearchServiceTest.java
package com.visor.school.search.service;

import com.visor.school.search.integration.SearchIndexClient;
import com.visor.school.search.model.SearchIndex;
import com.visor.school.search.model.SearchType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class SearchServiceTest {

    @Mock
    private SearchIndexClient searchIndexClient;

    @InjectMocks
    private SearchService searchService;

    @BeforeEach
    void setup() {
        searchService = new SearchService(searchIndexClient);
    }

    @Test
    void shouldIndexDocument() {
        // Given
        SearchIndex searchIndex = SearchIndex.builder()
                .id(UUID.randomUUID())
                .type(SearchType.STUDENT)
                .title("John Doe")
                .content("Student in Grade 10")
                .metadata(Map.of("gradeLevel", "10", "studentId", "STU-001"))
                .build();
        doNothing().when(searchIndexClient).index(searchIndex);

        // When
        searchService.index(searchIndex);

        // Then
        verify(searchIndexClient).index(searchIndex);
    }

    @Test
    void shouldSearchDocuments() {
        // Given
        String query = "John Doe";
        SearchType type = SearchType.STUDENT;
        List<SearchIndex> results = Collections.singletonList(
                SearchIndex.builder()
                        .id(UUID.randomUUID())
                        .type(SearchType.STUDENT)
                        .title("John Doe")
                        .content("Student in Grade 10")
                        .metadata(Map.of("gradeLevel", "10"))
                        .build());
        when(searchIndexClient.search(query, type)).thenReturn(results);

        // When
        List<SearchIndex> found = searchService.search(query, type);

        // Then
        assertEquals(1, found.size());
        assertEquals("John Doe", found.get(0).getTitle());
        verify(searchIndexClient).search(query, type);
    }

    @Test
    void shouldUpdateIndex() {
        // Given
        SearchIndex searchIndex = SearchIndex.builder()
                .id(UUID.randomUUID())
                .type(SearchType.STUDENT)
                .title("John Doe Updated")
                .content("Student in Grade 11")
                .metadata(Map.of("gradeLevel", "11"))
                .build();
        doNothing().when(searchIndexClient).updateIndex(searchIndex);

        // When
        searchService.updateIndex(searchIndex);

        // Then
        verify(searchIndexClient).updateIndex(searchIndex);
    }

    @Test
    void shouldDeleteIndex() {
        // Given
        UUID id = UUID.randomUUID();
        SearchType type = SearchType.STUDENT;
        doNothing().when(searchIndexClient).deleteIndex(id, type);

        // When
        searchService.deleteIndex(id, type);

        // Then
        verify(searchIndexClient).deleteIndex(id, type);
    }
}

File: services/search-service/src/test/kotlin/com/visor/school/search/service/SearchServiceTest.kt
package com.visor.school.search.service

import com.visor.school.search.integration.ElasticsearchClient
import com.visor.school.search.model.SearchIndex
import com.visor.school.search.model.SearchType
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class SearchServiceTest {

    @Mock
    private lateinit var elasticsearchClient: ElasticsearchClient

    @InjectMocks
    private lateinit var searchService: SearchService

    @BeforeEach
    fun setup() {
        searchService = SearchService(elasticsearchClient)
    }

    @Test
    fun `should index document`() {
        // Given
        val searchIndex = SearchIndex(
            id = UUID.randomUUID(),
            type = SearchType.STUDENT,
            title = "John Doe",
            content = "Student in Grade 10",
            metadata = mapOf("gradeLevel" to "10", "studentId" to "STU-001")
        )
        doNothing().whenever(elasticsearchClient).index(searchIndex)

        // When
        searchService.index(searchIndex)

        // Then
        verify(elasticsearchClient).index(searchIndex)
    }

    @Test
    fun `should search documents`() {
        // Given
        val query = "John Doe"
        val type = SearchType.STUDENT
        val results = listOf(
            SearchIndex(
                id = UUID.randomUUID(),
                type = SearchType.STUDENT,
                title = "John Doe",
                content = "Student in Grade 10",
                metadata = mapOf("gradeLevel" to "10")
            )
        )
        whenever(elasticsearchClient.search(query, type)).thenReturn(results)

        // When
        val found = searchService.search(query, type)

        // Then
        assertEquals(1, found.size)
        assertEquals("John Doe", found[0].title)
        verify(elasticsearchClient).search(query, type)
    }

    @Test
    fun `should update index`() {
        // Given
        val searchIndex = SearchIndex(
            id = UUID.randomUUID(),
            type = SearchType.STUDENT,
            title = "John Doe Updated",
            content = "Student in Grade 11",
            metadata = mapOf("gradeLevel" to "11")
        )
        doNothing().whenever(elasticsearchClient).updateIndex(searchIndex)

        // When
        searchService.updateIndex(searchIndex)

        // Then
        verify(elasticsearchClient).updateIndex(searchIndex)
    }

    @Test
    fun `should delete index`() {
        // Given
        val id = UUID.randomUUID()
        val type = SearchType.STUDENT
        doNothing().whenever(elasticsearchClient).deleteIndex(id, type)

        // When
        searchService.deleteIndex(id, type)

        // Then
        verify(elasticsearchClient).deleteIndex(id, type)
    }
}

